include/grpcpp/impl/codegen/server_callback.h

```diff
@@ -117,79 +191,575 @@ class CallbackUnaryHandler : public MethodHandler {
                                     std::function<void()> call_requester)
         : ctx_(ctx),
           call_(*call),
           req_(req),
-          call_requester_(std::move(call_requester)) {}
+          call_requester_(std::move(call_requester)) {
+      ctx_->BeginCompletionOp(call, [this](bool) { MaybeDone(); });
+    }
 
     ~ServerCallbackRpcControllerImpl() { req_->~RequestType(); }
 
-    RequestType* request() { return req_; }
+    const RequestType* request() { return req_; }
     ResponseType* response() { return &resp_; }
 
-    CallOpSet<CallOpSendInitialMetadata> meta_buf_;
+    void MaybeDone() {
+      if (--callbacks_outstanding_ == 0) {
+        grpc_call* call = call_.call();
+        auto call_requester = std::move(call_requester_);
+        this->~ServerCallbackRpcControllerImpl();  // explicitly call destructor
+        g_core_codegen_interface->grpc_call_unref(call);
+        call_requester();
+      }
+    }
+
+    CallOpSet<CallOpSendInitialMetadata> meta_ops_;
     CallbackWithSuccessTag meta_tag_;
     CallOpSet<CallOpSendInitialMetadata, CallOpSendMessage,
               CallOpServerSendStatus>
-        finish_buf_;
+        finish_ops_;
     CallbackWithSuccessTag finish_tag_;
 
     ServerContext* ctx_;
     Call call_;
-    RequestType* req_;
+    const RequestType* req_;
     ResponseType resp_;
     std::function<void()> call_requester_;
+    std::atomic_int callbacks_outstanding_{
+        2};  // reserve for Finish and CompletionOp
+  };
+};
+
+template <class RequestType, class ResponseType>
+class CallbackClientStreamingHandler : public MethodHandler {
+ public:
+  CallbackClientStreamingHandler(
+      std::function<void(ServerContext*, ResponseType*,
+                         experimental::ServerCallbackReader<RequestType>*)>
+          func)
+      : func_(func) {}
+  void RunHandler(const HandlerParameter& param) final {
+    // Arena allocate a reader structure (that includes response)
+    g_core_codegen_interface->grpc_call_ref(param.call->call());
+    auto* reader = new (g_core_codegen_interface->grpc_call_arena_alloc(
+        param.call->call(), sizeof(ServerCallbackReaderImpl)))
+        ServerCallbackReaderImpl(param.server_context, param.call,
+                                 std::move(param.call_requester));
+    Status status = param.status;
+
+    if (status.ok()) {
+      // Call the actual function handler and expect the user to call finish
+      CatchingCallback(func_, param.server_context, reader->response(), reader);
+    } else {
+      // if deserialization failed, we need to fail the call
+      reader->Finish(status);
+    }
+  }
+
+ private:
+  std::function<void(ServerContext*, ResponseType*,
+                     experimental::ServerCallbackReader<RequestType>*)>
+      func_;
+
+  class ServerCallbackReaderImpl
+      : public experimental::ServerCallbackReader<RequestType> {
+   public:
+    void BindReactor(experimental::ServerReadReactor* reactor) override {
```

Comment:
Is there a valid use case for re binding? If not, can we enforce that it is never re set?

Reply:
I'm thinking that there isn't so I'm ok with asserting. We might go a different direction here.